### PR TITLE
Fix setState calls, allow Promise factories

### PR DESCRIPTION
**Pull request by conartist6**
Calling setState on an unmounted component is an error! This is the kind of boilerplate (and error) that this package should be able to save users from writing (and causing).

Also allow a factory function to be passed in place of the current configuration, which if specified should be supplied with props and return a configuration object in the existing format.

### CODE REVIEW

**Comment by conartist6** [comment]
Hang on, self review, the code needs to handle what happens if the props change now.

**Comment by conartist6** [comment]
The newest code needs some additional tests, but I'm going to hold off on writing those until I hear some feedback and whether you're interested in merging this.

**Comment by matthis-d** [comment]
Hi @conartist6 !
Thanks a lot for your interest in my project. I was in vacation and didn't see all you did. Sorry about that. 

I'm reviewing it now ;)

File: README.md
Comment (matthis-d) [inline README.md]: I think you missed the `function` keyword or to use the fat-arrow like
```js
withPromises((props, oldProps, oldPromises) => ({ /* the promises object */ })(MyComponent)
```

**Review by matthis-d — APPROVED** [review]
Thanks about the code that avoids calling setState while the component did unmount.

**Comment by matthis-d** [comment]
I really like the idea of using a function instead of a simple object! 
I did think about it but now I'm reading it, it seems more natural, espacially when we are used to code with Redux.

File: README.md
Comment (conartist6) [reply README.md]: oops...

Comment (matthis-d) [reply README.md]: Do you prefer to do the modification or that I do it after merging your PR?

**Comment by matthis-d** [comment]
Thanks a lot ;)

**Comment by conartist6** [comment]
Thank you!

**Pull request merged**
